Remove supports_named_tensor from codegen entirely.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38806 Don't generate DeviceGuard for CPU wrapping code.
* **#38739 Remove supports_named_tensor from codegen entirely.**

Instead of codegenning the named tensor support checks into
CPUType/CUDAType, we instead add a new dispatch key that is put
into tensor whenever it has names.  By default, the fallback
implementation says that named tensors are not supported, but
if they are supported, we register a fallthrough which lets
us through to the true backend implementation.

There are a bunch of small pieces which are necessary to make this
happen:

- NameMode now also excludes DispatchKey::Named from the dispatch set
- To avoid bad error messages, we add a teensy special case to
  the dispatcher for named_not_supported_kernel: if we see that
  the boxed kernel we need to invoke from unboxed is this kernel,
  but we don't support boxing, but it's a kernel which is known
  to not need boxing, we just pass in nullptr for the stack.
  The special case here is very nice: it doesn't affect the fast
  path and only gets exercised when things are not supported.
- I need to add support for per operator fallthrough registration.
  This is done similarly to how we support fallthrough fallback,
  by just keeping track if the registered kernel for an operator
  is a fallthrough.

It is possible we could go even further down this path, and move
the named tensor logic itself into this key.  I leave this
up to future work.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21662643](https://our.internmc.facebook.com/intern/diff/D21662643)

ezyang committed:
Remove supports_named_tensor from codegen entirely.

Instead of codegenning the named tensor support checks into
CPUType/CUDAType, we instead add a new dispatch key that is put
into tensor whenever it has names.  By default, the fallback
implementation says that named tensors are not supported, but
if they are supported, we register a fallthrough which lets
us through to the true backend implementation.

There are a bunch of small pieces which are necessary to make this
happen:

- NameMode now also excludes DispatchKey::Named from the dispatch set
- To avoid bad error messages, we add a teensy special case to
  the dispatcher for named_not_supported_kernel: if we see that
  the boxed kernel we need to invoke from unboxed is this kernel,
  but we don't support boxing, but it's a kernel which is known
  to not need boxing, we just pass in nullptr for the stack.
  The special case here is very nice: it doesn't affect the fast
  path and only gets exercised when things are not supported.
- I need to add support for per operator fallthrough registration.
  This is done similarly to how we support fallthrough fallback,
  by just keeping track if the registered kernel for an operator
  is a fallthrough.

It is possible we could go even further down this path, and move
the named tensor logic itself into this key.  I leave this
up to future work.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Remove supports_named_tensor from codegen entirely."


Instead of codegenning the named tensor support checks into
CPUType/CUDAType, we instead add a new dispatch key that is put
into tensor whenever it has names.  By default, the fallback
implementation says that named tensors are not supported, but
if they are supported, we register a fallthrough which lets
us through to the true backend implementation.

There are a bunch of small pieces which are necessary to make this
happen:

- NameMode now also excludes DispatchKey::Named from the dispatch set
- To avoid bad error messages, we add a teensy special case to
  the dispatcher for named_not_supported_kernel: if we see that
  the boxed kernel we need to invoke from unboxed is this kernel,
  but we don't support boxing, but it's a kernel which is known
  to not need boxing, we just pass in nullptr for the stack.
  The special case here is very nice: it doesn't affect the fast
  path and only gets exercised when things are not supported.
- I need to add support for per operator fallthrough registration.
  This is done similarly to how we support fallthrough fallback,
  by just keeping track if the registered kernel for an operator
  is a fallthrough.

It is possible we could go even further down this path, and move
the named tensor logic itself into this key.  I leave this
up to future work.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Remove supports_named_tensor from codegen entirely."


Instead of codegenning the named tensor support checks into
CPUType/CUDAType, we instead add a new dispatch key that is put
into tensor whenever it has names.  By default, the fallback
implementation says that named tensors are not supported, but
if they are supported, we register a fallthrough which lets
us through to the true backend implementation.

There are a bunch of small pieces which are necessary to make this
happen:

- NameMode now also excludes DispatchKey::Named from the dispatch set
- To avoid bad error messages, we add a teensy special case to
  the dispatcher for named_not_supported_kernel: if we see that
  the boxed kernel we need to invoke from unboxed is this kernel,
  but we don't support boxing, but it's a kernel which is known
  to not need boxing, we just pass in nullptr for the stack.
  The special case here is very nice: it doesn't affect the fast
  path and only gets exercised when things are not supported.
- I need to add support for per operator fallthrough registration.
  This is done similarly to how we support fallthrough fallback,
  by just keeping track if the registered kernel for an operator
  is a fallthrough.

It is possible we could go even further down this path, and move
the named tensor logic itself into this key.  I leave this
up to future work.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/core/NamedRegistrations.cpp:
```diff
@@ -0,0 +1,458 @@
+  m.impl("backward", CppFunction::makeFallthrough());
+  m.impl("set_data", CppFunction::makeFallthrough());
+  m.impl("data", CppFunction::makeFallthrough());
+  m.impl("is_leaf", CppFunction::makeFallthrough());
```
Does it matter what the second argument (`CppFunction::makeFallthrough()`) is here? A lot of the autograd things have `manual_kernel_registration: True` in native_functions.yaml; from experimentation it looks like no matter whatever function is registered, the catch-all kernel gets called: https://github.com/pytorch/pytorch/blob/c60daedb362a2c93cb5c56a1a0c0851a6388b06d/torch/csrc/autograd/VariableTypeManual.cpp#L398

ezyang replied:
Yes, I believe it matters. The sequence of events that happens on `is_leaf` is that we first invoke at Variable dispatch key (since that is enabled by default and is first in priority), and then when we redispatch with Variable excluded, the next in line would be Named, and NOW we will hit this fallthrough case and keep going to CPU.

zou3519 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -78,7 +78,6 @@
   use_c10_dispatcher: full
   manual_kernel_registration: True
   variants: method
-  supports_named_tensor: True
```
Could you also delete https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/README.md#supports_named_tensor ?

ezyang commented on aten/src/ATen/core/boxing/KernelFunction.h:
```diff
@@ -17,6 +17,15 @@ struct OperatorKernel;
+// cased in the dispatcher to be triggered before we attempt boxing (so we can
+// give a good error message in cases when boxing is not supported).  When
+// boxing is universally supported this can be removed.
+[[noreturn]] CAFFE2_API void named_not_supported_kernel(OperatorKernel*, const OperatorHandle&, Stack*);
```
Yes, this sounds reasonable for vmap too. Hope we don't end up with too many of these before full boxing support comes though!!

ezyang commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -120,6 +117,11 @@ def TypedDict(name, attrs, total=True):  # type: ignore
 );
 """)
 
+NAMED_SUPPORTED_REGISTRATION = CodeTemplate("""\
```
Nope, should be deleted

ezyang commented on test/test_namedtensor.py:
```diff
@@ -1092,7 +1092,7 @@ def test_unflatten(self):
         named = torch.randn(3, 3, names=('N', 'C'))
         with self.assertRaisesRegex(
-                RuntimeError, "pdist is not yet supported with named tensors"):
+                RuntimeError, r"pdist.+is not yet supported with named tensors"):
```
```
        with self.assertRaisesRegex(
                RuntimeError, r"as_strided_.+is not yet supported with named tensors"):
            named.as_strided_((3, 3), (3, 1))
```

ezyang committed:
Update on "Remove supports_named_tensor from codegen entirely."


Instead of codegenning the named tensor support checks into
CPUType/CUDAType, we instead add a new dispatch key that is put
into tensor whenever it has names.  By default, the fallback
implementation says that named tensors are not supported, but
if they are supported, we register a fallthrough which lets
us through to the true backend implementation.

There are a bunch of small pieces which are necessary to make this
happen:

- NameMode now also excludes DispatchKey::Named from the dispatch set
- To avoid bad error messages, we add a teensy special case to
  the dispatcher for named_not_supported_kernel: if we see that
  the boxed kernel we need to invoke from unboxed is this kernel,
  but we don't support boxing, but it's a kernel which is known
  to not need boxing, we just pass in nullptr for the stack.
  The special case here is very nice: it doesn't affect the fast
  path and only gets exercised when things are not supported.
- I need to add support for per operator fallthrough registration.
  This is done similarly to how we support fallthrough fallback,
  by just keeping track if the registered kernel for an operator
  is a fallthrough.

It is possible we could go even further down this path, and move
the named tensor logic itself into this key.  I leave this
up to future work.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21662643](https://our.internmc.facebook.com/intern/diff/D21662643)

[ghstack-poisoned]

zou3519 reviewed (dismissed):
the named tensor parts and codegen of this LGTM, but I don't have context over the dispatcher pieces so am deferring to @bhosmer and @smessmer on review for those

ezyang commented:
Thanks Richard! I removed you from the reviewer list; waiting on smessmer/bhosmer

bhosmer commented on aten/src/ATen/core/boxing/KernelFunction.cpp:
```diff
@@ -18,6 +19,16 @@ void fallthrough_kernel(OperatorKernel*, const OperatorHandle&, Stack*) {
+void named_not_supported_kernel(OperatorKernel*, const OperatorHandle& op, Stack*) {
+  // DO NOT LOOK AT STACK, YOU HAVE SHORT CIRCUITED BOXING
+  // See Note [named_not_supported_kernel]
+  TORCH_CHECK(0,
```
Just curious, why 0 instead of false?

ezyang replied:
No good reason. I probably was copying something nearby.

bhosmer reviewed (approved):
LG, looking forward to yanking the special case stuff once boxing coverage is 100%

ezyang committed:
Update on "Remove supports_named_tensor from codegen entirely."


Instead of codegenning the named tensor support checks into
CPUType/CUDAType, we instead add a new dispatch key that is put
into tensor whenever it has names.  By default, the fallback
implementation says that named tensors are not supported, but
if they are supported, we register a fallthrough which lets
us through to the true backend implementation.

There are a bunch of small pieces which are necessary to make this
happen:

- NameMode now also excludes DispatchKey::Named from the dispatch set
- To avoid bad error messages, we add a teensy special case to
  the dispatcher for named_not_supported_kernel: if we see that
  the boxed kernel we need to invoke from unboxed is this kernel,
  but we don't support boxing, but it's a kernel which is known
  to not need boxing, we just pass in nullptr for the stack.
  The special case here is very nice: it doesn't affect the fast
  path and only gets exercised when things are not supported.
- I need to add support for per operator fallthrough registration.
  This is done similarly to how we support fallthrough fallback,
  by just keeping track if the registered kernel for an operator
  is a fallthrough.

It is possible we could go even further down this path, and move
the named tensor logic itself into this key.  I leave this
up to future work.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21662643](https://our.internmc.facebook.com/intern/diff/D21662643)

[ghstack-poisoned]

ezyang committed:
Update on "Remove supports_named_tensor from codegen entirely."


Instead of codegenning the named tensor support checks into
CPUType/CUDAType, we instead add a new dispatch key that is put
into tensor whenever it has names.  By default, the fallback
implementation says that named tensors are not supported, but
if they are supported, we register a fallthrough which lets
us through to the true backend implementation.

There are a bunch of small pieces which are necessary to make this
happen:

- NameMode now also excludes DispatchKey::Named from the dispatch set
- To avoid bad error messages, we add a teensy special case to
  the dispatcher for named_not_supported_kernel: if we see that
  the boxed kernel we need to invoke from unboxed is this kernel,
  but we don't support boxing, but it's a kernel which is known
  to not need boxing, we just pass in nullptr for the stack.
  The special case here is very nice: it doesn't affect the fast
  path and only gets exercised when things are not supported.
- I need to add support for per operator fallthrough registration.
  This is done similarly to how we support fallthrough fallback,
  by just keeping track if the registered kernel for an operator
  is a fallthrough.

It is possible we could go even further down this path, and move
the named tensor logic itself into this key.  I leave this
up to future work.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21662643](https://our.internmc.facebook.com/intern/diff/D21662643)

[ghstack-poisoned]

ezyang committed:
Update on "Remove supports_named_tensor from codegen entirely."


Instead of codegenning the named tensor support checks into
CPUType/CUDAType, we instead add a new dispatch key that is put
into tensor whenever it has names.  By default, the fallback
implementation says that named tensors are not supported, but
if they are supported, we register a fallthrough which lets
us through to the true backend implementation.

There are a bunch of small pieces which are necessary to make this
happen:

- NameMode now also excludes DispatchKey::Named from the dispatch set
- To avoid bad error messages, we add a teensy special case to
  the dispatcher for named_not_supported_kernel: if we see that
  the boxed kernel we need to invoke from unboxed is this kernel,
  but we don't support boxing, but it's a kernel which is known
  to not need boxing, we just pass in nullptr for the stack.
  The special case here is very nice: it doesn't affect the fast
  path and only gets exercised when things are not supported.
- I need to add support for per operator fallthrough registration.
  This is done similarly to how we support fallthrough fallback,
  by just keeping track if the registered kernel for an operator
  is a fallthrough.

It is possible we could go even further down this path, and move
the named tensor logic itself into this key.  I leave this
up to future work.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21662643](https://our.internmc.facebook.com/intern/diff/D21662643)

[ghstack-poisoned]
